### PR TITLE
Updated retry logic

### PR DESCRIPTION
**Pull request by PLNech**

### CODE REVIEW

**Comment by redox** [comment]
LGTM 👍

File: algoliasearch/src/test/java/com/algolia/search/saas/IndexTest.java [hunk 1]
```diff
@@ -1113,4 +1116,57 @@ public void doRequestCompleted(JSONObject content, AlgoliaException error) {
...
+
+
+    // Should not reconnect before delay to the same host twice if it fails
+    // Should reconnect after delay to the same host twice if it fails
+
+    private String getRandomString() {
+        String randomAppId;
+        byte[] array = new byte[7]; // length is bounded by 7
+        new Random().nextBytes(array);
+        randomAppId = new String(array, Charset.forName("UTF-8"));
```
Comment (clement-leprovost) [inline algoliasearch/src/test/java/com/algolia/search/saas/IndexTest.java]: - I would be really surprised if this call succeeded in all cases. Not every sequence of bytes is valid UTF-8.
- Besides, using arbitrary characters in URLs is probably not a good idea, even if now many parts of Unicode are accepted (like Chinese Han characters, for example).

To solve both issues, I would prefer using a hexadecimal representation.

File: algoliasearch/src/main/java/com/algolia/search/saas/AbstractClient.java [hunk 2]
```diff
@@ -275,12 +296,30 @@ public int getSearchTimeout() {
...
+        checkTimeout(searchTimeout);
         this.searchTimeout = searchTimeout;
     }
 
     /**
+     * Get the timeout for retrying connection to a down host.
+     *
+     * @return The delay before connecting again to a down host (ms).
+     */
+    public int getHostDownDelay() {
```
Comment (clement-leprovost) [inline algoliasearch/src/main/java/com/algolia/search/saas/AbstractClient.java]: I like the "host down delay" name, better than the one I used in Swift ("host status timeout"). The problem is the Swift version is already released... :/ Alright, just keep it, I will maybe one day introduce a new property and deprecate the old one.

File: algoliasearch/src/main/java/com/algolia/search/saas/AbstractClient.java [hunk 3]
```diff
@@ -87,6 +88,24 @@ public int hashCode() {
         }
     }
 
+    static class HostStatus {
+        public boolean isUp = true;
+        public long lastTryTimestamp;
+
+        HostStatus() {
```
Comment (clement-leprovost) [inline algoliasearch/src/main/java/com/algolia/search/saas/AbstractClient.java]: Looks like dead code (could not find any usage of this constructor).

File: algoliasearch/src/main/java/com/algolia/search/saas/AbstractClient.java [hunk 4]
```diff
@@ -87,6 +88,24 @@ public int hashCode() {
...
+            lastTryTimestamp = new Date().getTime();
+        }
+
+        HostStatus(boolean isUp) {
+            this.isUp = isUp;
+            lastTryTimestamp = new Date().getTime();
+        }
+
+        void updateState(boolean isUp) {
```
Comment (clement-leprovost) [inline algoliasearch/src/main/java/com/algolia/search/saas/AbstractClient.java]: Looks like dead code. `HostStatus` looks like an immutable class in spirit, anyway.

File: algoliasearch/src/test/java/com/algolia/search/saas/IndexTest.java (hunk 1 above)
Comment (PLNech) [inline algoliasearch/src/test/java/com/algolia/search/saas/IndexTest.java]: Fixed in 0814e9c by using `UUID#randomUUID`.

File: algoliasearch/src/main/java/com/algolia/search/saas/AbstractClient.java (hunk 3 above)
Comment (PLNech) [inline algoliasearch/src/main/java/com/algolia/search/saas/AbstractClient.java]: Fixed in 4647650.

File: algoliasearch/src/main/java/com/algolia/search/saas/AbstractClient.java (hunk 4 above)
Comment (PLNech) [inline algoliasearch/src/main/java/com/algolia/search/saas/AbstractClient.java]: Fixed in 4647650d0040f1b0f46366f866519c1af9cbd76f.

**Comment by clement-leprovost** [comment]
👍 LGTM

**Comment by PLNech** [comment]
Cancelled [job #702.4](https://travis-ci.org/algolia/algoliasearch-client-android/jobs/182307554) which stalled after the tests succeeded.

**Pull request merged**